pkg/resources/external_volume.go

```diff
@@ -0,0 +1,711 @@
+	}
+
+	if err := client.ExternalVolumes.Alter(ctx, sdk.NewAlterExternalVolumeRequest(id).WithAddStorageLocation(*newStorageLocationreq)); err != nil {
+		return err
+	} else {
+		return nil
+	}
+}
+
+func removeStorageLocation(
+	removedLocation sdk.ExternalVolumeStorageLocation,
+	client *sdk.Client,
+	ctx context.Context,
+	id sdk.AccountObjectIdentifier,
+) error {
+	removedName, err := GetStorageLocationName(removedLocation)
+	if err != nil {
+		return err
+	}
+
+	if err := client.ExternalVolumes.Alter(ctx, sdk.NewAlterExternalVolumeRequest(id).WithRemoveStorageLocation(removedName)); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Process the removal / addition storage location requests.
+// to avoid creating storage locations with duplicate names.
+// len(removedLocations) should be less than the total number
+// of storage locations the external volume has, else this function will fail.
+func updateStorageLocations(
+	removedLocations []sdk.ExternalVolumeStorageLocation,
+	addedLocations []sdk.ExternalVolumeStorageLocation,
+	client *sdk.Client,
+	ctx context.Context,
+	id sdk.AccountObjectIdentifier,
+) error {
+	for _, removedLocation := range removedLocations {
+		err := removeStorageLocation(removedLocation, client, ctx, id)
+		if err != nil {
+			return err
+		}
+	}
+	for _, addedLocation := range addedLocations {
+		err := addStorageLocation(addedLocation, client, ctx, id)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func UpdateContextExternalVolume(ctx context.Context, d *schema.ResourceData, meta any) diag.Diagnostics {
+	client := meta.(*provider.Context).Client
+	id, err := sdk.ParseAccountObjectIdentifier(d.Id())
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	set := sdk.NewAlterExternalVolumeSetRequest()
+
+	if d.HasChange("comment") {
+		// not using d.GetOk as that doesn't let comments be reset to the empty string
+		set.WithComment(d.Get("comment").(string))
+	}
+
+	if d.HasChange("allow_writes") {
+		if v := d.Get("allow_writes").(string); v != BooleanDefault {
+			parsed, err := booleanStringToBool(v)
+			if err != nil {
+				return diag.FromErr(err)
+			}
+			set.WithAllowWrites(parsed)
+		} else {
+			// no way to unset allow writes - set to false as a default
+			set.WithAllowWrites(false)
+		}
+	}
+
+	if (*set != sdk.AlterExternalVolumeSetRequest{}) {
+		if err := client.ExternalVolumes.Alter(ctx, sdk.NewAlterExternalVolumeRequest(id).WithSet(*set)); err != nil {
+			return diag.FromErr(err)
+		}
+	}
+
+	if d.HasChange("storage_location") {
+		old, new := d.GetChange("storage_location")
+		oldLocations, err := extractStorageLocations(old, false)
+		if err != nil {
+			return diag.FromErr(err)
+		}
+
+		newLocations, err := extractStorageLocations(new, false)
+		if err != nil {
+			return diag.FromErr(err)
+		}
+
+		// Storage locations can only be added to the tail of the list, but can be
```

Comment:
Have gone to the trouble of adding/removing storage locations here to avoid forcing new on the volume, as other objects will depend on it (e.g. iceberg tables). Implementation would be simpler if we simply rebuilt the entire storage location list each update, but that obviously has the potential to produce many more Snowflake queries.